Added BlindWalk potion for Google Code In 2016

arpitkamboj opened the pull request:
Added BlindWalk potion - Turns the player blind for 5 seconds along with speed decrease.

*Cervator edit: tied to GCI task https://codein.withgoogle.com/dashboard/task-instances/4855239574290432/*

arpitkamboj committed:
Added BlindWalk potion.

Added BlindWalk potion - Turns the player blind for 5 seconds along with speed decrease.

arpitkamboj committed:
Added BlindWalk potion.

Added BlindWalk potion - Turns the player blind for 5 seconds along with speed decrease.

arpitkamboj committed:
Added BlindWalk potion.

Added BlindWalk potion -Turns the player blind for 5 seconds along with speed decrease.

rzats commented on assets/prefabs/potion/BlindWalkPotion.prefab:
```diff
@@ -0,0 +1,33 @@
+                "magnitude": 0.5,
+                "duration": 5100
+            },
+			{
```
Also, the indentation here is a little odd - can you reformat the JSON file using IntelliJ?

arpitkamboj replied:
Thanks for the advice... I was a little confused at first on how all the effects were implemented.... Now its all clear... Thanks. Will update soon..

rzats reviewed (changes requested):
You'll actually have to implement the `NO_VISIBILITY` effect 🙂 For this you'll need to declare it in [`PotionCommonEffects`](https://github.com/Terasology/Potions/blob/master/src/main/java/org/terasology/potions/PotionCommonEffects.java), then implement in [`DrinkPotionAuthoritySystem`](https://github.com/Terasology/Potions/blob/master/src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java).

arpitkamboj commented:
Hey,
Even if I implement the 'No-Visibility' effect in common effects and potionauthoritysystem, there is no code to specify what the game has to do...

Please help

rzats commented:
That's correct - actually implementing the code is also part of the task 🙂

arpitkamboj commented:
Can you please give some idea on how to do that?

rzats commented:
I've done some research on this and found out that making the effect work requires an API that's not yet merged (https://github.com/MovingBlocks/Terasology/pull/2449). Therefore, it's OK for you to submit a stub method/code block with a TODO note for later 🙂

arpitkamboj commented:
Thanks.
I'll do that for now.
But I figured out that I need to create an entire new file for the No_visibility effect and set it to render a black screen for the given amount of time. I will work on this later.

-Arpit

arpitkamboj committed:
Added BlindWalk potion.

Added BlindWalk potion. Waiting for an API to be merged that would allow to make No_Visibility effect actually happen.

arpitkamboj committed:
Added BlindWalk potion.

Added BlindWalk potion. Waiting for an API to be merged that would allow to make No_Visibility effect actually happen.

arpitkamboj committed:
Added BlindWalk potion.

Added BlindWalk potion - Turns the player blind for 5 seconds along with speed decrease.

Cervator commented on assets/prefabs/potion/BlindWalkPotion.prefab:
```diff
@@ -0,0 +1,33 @@
+                "duration": 5100
+            },
+            {
+		"effect": "NO_VISIBILITY"
```
A few stray tab characters here, please replace with spaces :)

Cervator commented on src/main/java/org/terasology/potions/PotionCommonEffects.java:
```diff
@@ -44,6 +44,7 @@
     public static final String FEATHER_FALL = "FEATHER_FALL";
     public static final String BRICK_FALL = "BRICK_FALL";
     public static final String MULTI_JUMP = "MULTI_JUMP";
+	public static final String NO_VISIBILITY = "NO_VISIBILITY";
```
Could you add something above this to indicate the pending effect implementation? Like:

`// TODO: Effect implementation awaiting API improvements, such as https://github.com/MovingBlocks/Terasology/pull/2449

Cervator commented on src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java:
```diff
@@ -217,6 +217,9 @@ public void onPotionWithoutGenomeConsumed(DrinkPotionEvent event, EntityRef ref)
                     herbEffect = new ExplosiveEffect();
                     break;
+				case PotionCommonEffects.NO_VISIBILITY:
+                    herbEffect = new NoVisibilityEffect();
```
Note that this class does not appear to be part of the PR. Forget to `git add` it? :-)

arpitkamboj replied:
OK I get it I need to add a file for NoVisibilityEffect to the potions/effects directory....

Thanks

Cervator replied:
You have the class now, but the DrinkPotionAuthoritySystem needs to import it before it'll be able to find it. Again something like IntelliJ should outright prompt you "Hey want to import this class from X?" :-) 

You also have tab characters again on the case statement line

Cervator reviewed (changes requested):
Heyo, few more tiny tweaks requested, although agreed on the effect implementation being problematic right at this moment so we can merge and approve in the meantime :-)

One additional note: the text on the potion icon will be *very* hard to see in-game (and even get reversed due to the way held items are rendered right now). Maybe just pick a color or simple icon? Maybe an outlined footprint (dashed line) instead

Cervator commented:
And just to echo @rzats yeah it'll be fine to submit a more or less blank stub class for `NoVisibilityEffect` - we just need the code to actually compile before we can merge it :D

arpitkamboj committed:
Update BlindWalkPotion.prefab

arpitkamboj committed:
Update BlindWalkPotion.prefab

arpitkamboj committed:
Update BlindWalkPotion.prefab

arpitkamboj committed:
Added BlindWalk potion.

arpitkamboj committed:
Added BlindWalk potion.

Added the NoVisibility Effect file for the BlindWalk that will make the screen go blank for a given amount of time.
The effect will be worked upon, once the API for custom actions is implemented.

arpitkamboj committed:
Update README.md

arpitkamboj committed:
Update README.md

arpitkamboj committed:
Update README.md

arpitkamboj committed:
Update README.md

arpitkamboj committed:
Added BlindWalk potion

Added a comment referring to the API implementation issue.

arpitkamboj commented:
Hey,
Please review the changes I made to the files.

Thanks
-Arpit

arpitkamboj committed:
Added BlindWalk potion

Cervator commented on src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:
```diff
@@ -0,0 +1,42 @@
+ */
+package org.terasology.potions.effect;
+
+import org.terasology.entitySystem.entity.EntityManager;
```
Most of these imports are unused, should show gray and underlined in IntelliJ - please remove those :-)

Cervator commented on src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:
```diff
@@ -0,0 +1,42 @@
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.logic.actions.*;
```
Wildcard imports are against our code style, although these will be removed anyway as unused. But if you configure IntelliJ right it'll avoid the wildcard imports / format them nicely in the future :-)

See http://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports

Cervator commented on src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:
```diff
@@ -0,0 +1,42 @@
+ * Once the API for Custom Actions is implemented, the code for NoVisibility Effect goes here.
+ The effect will make the screen blank for the given amount of time.
+ */
+public class NoVisibilityEffect implements HerbEffect {
```
If you were compiling the code to make sure all was well you should've seen an error here. Like other effects even though this one is a placeholder pending API improvements anything implementing `HerbEffect` also must implement the `applyEffect` potion. Check out other effects to see how it should look - yours can probably be empty but needs to exist. It looks like you've got the Javadoc for it below, so maybe you just deleted the method header itself?

Cervator commented on src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:
```diff
@@ -0,0 +1,42 @@
+     * @param magnitude     The magnitude of the effect.
+     * @param duration      The duration of the effect in milliseconds.
+     
+	 Once the API for Custom Actions is implemented, the code for NoVisibility Effect goes here.
```
You have a couple tab characters here. Please replace with spaces :-)

Cervator commented on README.md:
```diff
@@ -146,4 +151,6 @@ Send in PRs to this repository and feel free to add your name to the authors lis
-[Trekmarvel](https://github.com/Trekmarvel) for the base potion bottle graphics.
+* [Trekmarvel](https://github.com/Trekmarvel) for the base potion bottle graphics.
+* InvinciblePotionBottle.png, Regen2PotionBottle.png edited from base by Minege
+* MegaJumpBottle.png, GigaJumpBottle.png edited from base by Patrick Wang
```
After this line you could put a quick note that "BlindWalkPotionBottle.png edited from base by digitalripperynr"

Cervator reviewed (changes requested):
So at this point there's more small stuff that would be super easy to avoid if you were running out of IntelliJ and noticing its little recommendations. If you are at all able to get IntelliJ running with the source code showing in there that would be the best thing you could do right now :-)

Each issue is super easy to fix and I considered during it as part of the merge, but you really need to have a good setup to move forward with your coding adventure. Remember that IRC is available if you need help setting up!

arpitkamboj commented:
Hey,
I apologise. I messed up everything.
Now installing IntelliJ.
I will close this PR and open a new one with all the changes (final this time).

Thanks
-Arpit

arpitkamboj commented:
I am closing this PR and opening a new one...
